### ISSUE TITLE
Add dependencies versions and manual build instructions to documentation - WatchDog Tiggered when built from Source

### ISSUE THREAD

**Issue by anettleship** [interactionresearchstudio/ESP32-SOCKETIO#63]
I'm trying to build and upload the project to an esp32 using Arduino IDE and I can't find anything currently in the documentation that describes how to do this.

I would like to contribute by adding a section to the readme on how to build the project from source.

I've added the dependencies listed to my Arduino library and flashed the data directory to the esp32's storage. 

When I connect to the captive portal, the esp32 reboots when i try to select my network, at the point that I select the network drop down list. 

Clearly I've done something wrong, so the first thing I'd like to do is check which versions of the project dependencies I need, and any other steps I'm missing. 

Can you tell me which versions I need please? Once I've built successfully, I'll create a pull request with updated docs. 

Thanks

**Comment by anettleship** [comment]
Here's the procedure I've followed so far using Arduino IDE:

- Git clone the ESP32-SOCKETIO repository from github to a local folder
- Locate the Arduino Libraries Folder and git clone all dependencies listed in the readme file to this location:
- e.g. C:\Users\Username\Documents\Arduino\Libraries

![image](https://github.com/interactionresearchstudio/ESP32-SOCKETIO/assets/5358837/7abc3528-3706-4d0f-8ab3-cab98c2f46cc)

* [AsyncTCP](https://github.com/me-no-dev/AsyncTCP)
* [ESPAsyncWebServer](https://github.com/me-no-dev/ESPAsyncWebServer)
* [SocketIOClient](https://github.com/timum-viw/socket.io-client)
* [ArduinoWebSockets](https://github.com/Links2004/arduinoWebSockets)
* [ArduinoJSON](https://github.com/bblanchon/ArduinoJson)
* [AceButton](https://github.com/bxparks/AceButton)
* [FastLed](https://github.com/FastLED/FastLED) (Light Touch)
* [ESP32Servo](https://madhephaestus.github.io/ESP32Servo/annotated.html) (Speed Dial)

- Compile and upload the project to your ESP32 using Arduino IDE
- Using the instructions from this project, download and install the esp32fs plugin. This requires us to use Arduino IDE v1.x - the plug in is not compatible with Arduino IDE v2+ and the menu Tools > ESP32 Sketch Data Upload menu item does not appear!
- https://github.com/me-no-dev/arduino-esp32fs-plugin
- Use the ESP32 Sketch Data Upload menu from Arduino IDE v1.x to upload the data folder to the ESP32 fs storage - this enables the device to serve the html and js files required to display the captive portal.

Connect to the YO-YO device wifi and follow standard set up procedure.

**Comment by anettleship** [comment]
Currently, the device reboots when I access the captive portal. This appears to co-incide with the call to /scan.

```
handleFileRead: /index.html
handleRequest: /generate_204
handleFileRead: /index.html
handleRequest: /bootstrap-4.4.1.min.css
handleFileRead: /bootstrap-4.4.1.min.css
handleRequest: /style.css
handleFileRead: /style.css
handleRequest: /jquery-3.5.1.min.js
handleFileRead: /jquery-3.5.1.min.js
handleRequest: /bootstrap-4.4.1.min.js
handleFileRead: /bootstrap-4.4.1.min.js
handleRequest: /script.js
handleFileRead: /script.js
handleRequest: /logo_yoyo.png
handleFileRead: /logo_yoyo.png
handleRequest: /generate_204
handleFileRead: /index.html
handleRequest: /credentials
getCredentials
pairedSetup
handleRequest: /favicon.ico
handleRequest: /generate_204
handleFileRead: /index.html
handleRequest: /scan
E (43774) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (43774) task_wdt:  - async_tcp (CPU 0/1)
E (43774) task_wdt: Tasks currently running:
E (43774) task_wdt: CPU 0: IDLE
E (43774) task_wdt: CPU 1: loopTask
E (43774) task_wdt: Aborting.

abort() was called at PC 0x400f3d08 on core 0


Backtrace: 0x40083a75:0x3ffbecbc |<-CORRUPTED




ELF file SHA256: a3d64e267fd50348

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13964
load:0x40080400,len:3600
entry 0x400805f0
Your fade time is 6 hours
Touch threshold is:45
Already has one or more paired mac address
Scanning for available SCADS
scan start
```

The debug messages mention async_tcp. The version that comes from the asyn_tcp repository is v1.1.1 but the Arduino IDE has v1.1.4 available. Updating to this version produces similar results:


```

handleFileRead: /index.html
handleRequest: /style.css
handleFileRead: /style.css
handleRequest: /jquery-3.5.1.min.js
handleFileRead: /jquery-3.5.1.min.js
handleRequest: /bootstrap-4.4.1.min.js
handleFileRead: /bootstrap-4.4.1.min.js
handleRequest: /script.js
handleFileRead: /script.js
handleRequest: /logo_yoyo.png
handleFileRead: /logo_yoyo.png
handleRequest: /credentials
getCredentials
pairedSetup
handleRequest: /favicon.ico
handleRequest: /generate_204
handleFileRead: /index.html
handleRequest: /scan
E (39583) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (39583) task_wdt:  - async_tcp (CPU 0/1)
E (39583) task_wdt: Tasks currently running:
E (39583) task_wdt: CPU 0: IDLE
E (39583) task_wdt: CPU 1: loopTask
E (39583) task_wdt: Aborting.

abort() was called at PC 0x400f3d08 on core 0


Backtrace: 0x40083a75:0x3ffbecbc |<-CORRUPTED




ELF file SHA256: a3d64e267fd50348

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13964
load:0x40080400,len:3600
entry 0x400805f0



```



A quick search suggests that the Watchdog is a mechanism to stop a single process locking out the CPU, that we're falling foul of. 


What's the step that I'm missing to compile a working image from source code that behaves like the images that are delivered by the YOYO firmware uploader?

**Comment by anettleship** [comment]
I thought I had it there... I had missed out the step to check out the latest tagged release, v0.5.8.

I've done that, and getting the same error still:

```
: /generate_204
handleFileRead: /index.html
handleRequest: /credentials
getCredentials
pairedSetup
handleRequest: /favicon.ico
handleRequest: /generate_204
handleFileRead: /index.html
handleRequest: /scan
E (45831) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (45831) task_wdt:  - async_tcp (CPU 0/1)
E (45831) task_wdt: Tasks currently running:
E (45831) task_wdt: CPU 0: IDLE
E (45831) task_wdt: CPU 1: loopTask
E (45831) task_wdt: Aborting.

abort() was called at PC 0x400f3f70 on core 0


Backtrace: 0x40083a75:0x3ffbecbc |<-CORRUPTED




ELF file SHA256: 49023d7d320590e0

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13964
load:0x40080400,len:3600
entry 0x400805f0

```

I'm going to set a next step to check out an earlier release, and modify the debug output slightly to show that I'm successfully flashing the code version I think I'm sending.

**Comment by anettleship** [comment]
One possible approach appears to be implementing the watchdog timer configuration in the code base. There appear to be ways to setting this up in the configuration when using different IDEs to compile the sketch, such as Platform IO, but I haven't fully understood this yet.

The code based approach is described in a couple of forms:
https://iotassistant.io/esp32/enable-hardware-watchdog-timer-esp32-arduino-ide/
https://stackoverflow.com/questions/66278271/task-watchdog-got-triggered-the-tasks-did-not-reset-the-watchdog-in-time

I'm able to add the watchdog config and set it on the main loop by first importing and setting the timer value:

```
#include <esp_task_wdt.h>
// 120 seconds WDT
#define WDT_TIMEOUT 120
```

Then adding this to the setup() function:

```
  // setup WDT
  Serial.println("Configuring WDT...");
  esp_task_wdt_init(WDT_TIMEOUT, true); //enable panic so ESP32 restarts
  esp_task_wdt_add(NULL); //add current thread to WDT watch
```

And then this to the loop function to 'feed' the watchdog - i.e. tell it that the process it is watching is still alive, and thus prevent it from resetting our device in order to recover from a crash / hang / infinite loop / hardware failure:

```
  esp_task_wdt_reset();
```

When I access the Captive Portal, my ESP32 no longer resets. Light touch functionality appears to be working as expected.

There clearly is a way to set this configuration without modifying the codebase, but I am not able to find it. Can anyone guide me on this please?

**Comment by mikevanis** [comment]
I'm not sure I understand what you're trying to do with the watchdog. The process to build and upload from the Arduino IDE is as follows:

1. Download code
2. Install esp32 core
3. Install required libraries
4. Install esp32 sketch data upload [like this](https://randomnerdtutorials.com/install-esp32-filesystem-uploader-arduino-ide/)
5. Upload code
6. Upload data folder

We use the Arduino IDE for development, so there's no need to modify the codebase. I suspect the error you're getting is down to one of these reasons:
- You've selected the wrong board / SPIFFS configuration in the IDE
- You haven't uploaded the data folder

**Comment by anettleship** [comment]
Thanks, I'll admit it's a weird one, I've not encountered the watch dog before when working with esp8266 boards. 

I brought the watch dog code in to avoid the message quoted above:

Task watchdog got triggered 

...and subsequent reboot. 

I'll double check the board name I've chosen in the Arduino IDE matches my hardware. I'm using a pair of boards from pimoroni sold specifically with this project in mind (https://shop.pimoroni.com/products/yo-yo-machines-kit-1-light-touch) and I've chosen "ESP32 Dev Module" in the Arduino IDE and the SPIFFS configuration that's described in this page:

https://github.com/interactionresearchstudio/YoYoWiFiManager

"...for the ESP32 select a Partition Scheme of Default 4MB with spiffs (1.2MB APP/1.5MB SPIFFS)"

The data folder needs to be uploaded to the board using Arduino v1.x, I will try uploading the code using the same version - I've been using the latest version of v2.x to upload, perhaps this is the cause. 

Can you think of any other variables that might be different between your environment and mine?

Thanks again

**Comment by mikevanis** [comment]
Yes definitely use the v1.x IDE. We've never tried it with v2.

**Comment by anettleship** [comment]
&TLDR

Here are notes on how to follow the above procedure. Using Arduino IDE v1.8.19 for upload, my build still triggers the watchdog and the device reboots as described above. So although closed, the question of how to build this project from source in 2024 is still unresolved.

It is necessary to use:
v2.0.9 of esp32 core
v7.2.0 of ArduinoJson

All other dependencies are the current version as of 06/11/24.

---


Writing some more notes from an attempt to follow these instructions on 6th November 2024. 

I had to install an earlier version of esp32 core... current is v3.x.x which fails due to an issue with the current wifi.h that resulted in an error during build.

v2.0.14 caused an issue with my connection to the yoyo wifi network - I could not connect.
v1.06 would not build due to an error regarding wifi.h.

v2.0.9 appears to be the sweet spot - the build succeeds and I can connect to the yoyo wifi network.

(I'm aware the project also includes a wifi.h file, but to use this over the global version installed with esp32 core appears to require me to modify the codebase to specify the import path for wifi.h, which I do not wish to do).

![image](https://github.com/user-attachments/assets/afcabf0c-4e50-42ae-a109-3252732df9b8)

Having checkout out all the dependencies into a folder from their respective git repos, I had no issues using the latest versions of each as of this date.

I am playing with light touch, so I skipped over ESP32Servo (which has a confusing website instead of git hub page), and had no issues from doing so.

My build does still throw red warnings about ArduinoJSON, but the build succeeds. 

I checked out version 6.21.3 of ArduinoJson to see if I could get rid of the warning messages, the build failed.
I checked out version 6.20.0, the build failed.
Confirmed it passes by checking out tag v7.2.0.
